src/sources/prometheus/scrape.rs

```diff
@@ -0,0 +1,430 @@
+use super::parser;
+use crate::{
+    config::{self, GenerateConfig, GlobalOptions, SourceConfig, SourceDescription},
+    http::Auth,
+    http::HttpClient,
+    internal_events::{
+        PrometheusErrorResponse, PrometheusEventReceived, PrometheusHttpError,
+        PrometheusParseError, PrometheusRequestCompleted,
+    },
+    shutdown::ShutdownSignal,
+    sources,
+    tls::{TlsOptions, TlsSettings},
+    Event, Pipeline,
+};
+use futures::{compat::Sink01CompatExt, stream, FutureExt, StreamExt, TryFutureExt};
+use futures01::Sink;
+use hyper::{Body, Request};
+use serde::{Deserialize, Serialize};
+use snafu::ResultExt;
+use std::{
+    future::ready,
+    time::{Duration, Instant},
+};
+
+#[derive(Deserialize, Serialize, Clone, Debug)]
+struct PrometheusScrapeConfig {
+    // Deprecated name
+    #[serde(alias = "hosts")]
```

Comment:
This alias doesn't seem to be working. If I have

```
  hosts = ["http://localhost:9100/metrics"]
```

in my config, I get:

```
Nov 27 18:15:06.173 ERROR vector::cli: Configuration error. error=missing field `endpoints` for key `sources.node` at line 74 column 1

```

Reply:
This is sadly a [known serde bug](https://github.com/serde-rs/serde/issues/1504) caused by interaction between alias and flatten. This works when the source type is `prometheus_scrape`, but breaks for backwards compatible `prometheus`. I'm not sure what the best path to resolve this is, short of duplicating a lot of code, which makes the whole point of a simple alias somewhat moot.

Reply:
Ah, I didn't see there was a flatten behind it. Yeah, I've bumped into that bug before. Would you be able to define both fields in the struct and then validate in the loading code? I'm not sure how important it is that we maintain backward compatibility here?

Reply:
I guess I'll have to.

Reply:
If it helps, I feel your pain...